GPU/GPUCommon.cpp

```diff
@@ -2379,13 +2381,17 @@ void GPUCommon::Execute_ImmVertexAlphaPrim(u32 op, u32 diff) {
 		return;
 	}
 
+	int prim = (op >> 8) & 0x7;
+	if (prim != GE_PRIM_KEEP_PREVIOUS) {
+		// Flush before changing the prim type.  Only continue can be used to continue a prim.
+		FlushImm();
+	}
+
 	TransformedVertex &v = immBuffer_[immCount_++];
 
 	// Formula deduced from ThrillVille's clear.
-	int offsetX = gstate.getOffsetX16();
-	int offsetY = gstate.getOffsetY16();
-	v.x = ((gstate.imm_vscx & 0xFFFFFF) - offsetX) / 16.0f;
-	v.y = ((gstate.imm_vscy & 0xFFFFFF) - offsetY) / 16.0f;
+	v.x = ((gstate.imm_vscx & 0xFFFF) - 0x8000) / 16.0f;
+	v.y = ((gstate.imm_vscy & 0xFFFF) - 0x8000) / 16.0f;
```

Comment:
Gonna have to be

```
((int)(gstate.imm_vscx & 0xFFFF) - 0x8000) / 16.0f;
```
Otherwise we'll have an unsigned value after subtraction if the result is negative, which will be converted to floating point as a huge number.

Reply:
Are you sure that offsetX and offsetY are not involved? Without them, the rectangle in ThrillVill becomes -240, -136 to 240, 136 in through mode coordinates, instead of the expected and working 0,0 to 480,272.

Of course in ThrillVille'sc case they're just 0, so maybe there is no -0x8000 subtraction, or the offset comes from elsewhere...

Reply:
Oh right.  All my test verts were positive, oops, didn't think about negative hard... had just found the offset seemed fixed.

Well, in my test I tried varying the offset registers and the primitives did not translate.  To target a specific X I just had to send in the register: `((2048 + v[i].x) << 4)`.  Even `0x00FF0000 | ((2048 + v[i].x) << 4)` was the same.

This worked even when I used `sceGuViewport(0, 0, 1, 1); sceGuOffset(0, 0);` (which previously made graphics not draw at all in PPSSPP.)  The values of those registers make no difference to what gets drawn.

Can you attach a frame dump including ThrillVille's case?  Maybe there's some other register that affects it.

-[Unknown]

Reply:
[thrillville title ULES00889_0002.zip](https://github.com/hrydgard/ppsspp/files/9499548/thrillville.title.ULES00889_0002.zip)

Here's a framedump on the title screen, where it only clears a quarter of the screen.
Playing back the framedump crashes in FramebufferMaanger.cpp on line 813 for some reason, easily fixed by adding `currentRenderVfb_ != nullptr` to the condition (should probably commit that).

Reply:
Interesting.  I found some more issues looking at this dump, gonna dig into it a bit more in just a bit.  Pushed some updates (including two fixes for crashes.)

My immediate thought was maybe it was clear - which it isn't.  But clear does apply, prevents texturing, and masks rgb/s/z per its flags.  So that's good to have confirmed.  Lots of state that could be checked here, sigh.

What's also weird is softgpu is skipping the first draw (even in the TL quadrant), and I don't know why yet.

-[Unknown]

Reply:
Okay, that's fun.  It depends on the current vertextype register.  If that has throughmode set, then offset is fixed.  Otherwise, offset is applied.  This probably also explains why I couldn't get fog to apply.

-[Unknown]

Reply:
Alright, investigated the other flags (confirmed we can use this to directly set the 8-bit fog factor, which was actually my original goal investigating this, because the softgpu fog formula is still inaccurate and the floats got in the way.)

Here's how it compares now with that dump vs PSP (__testcompare output, transparent pixels highlight inaccuracies):
![__testcompare.png](https://user-images.githubusercontent.com/191233/188795098-16a251b1-bf9a-4f54-bc01-365275dad27f.png)

Still off in a bunch of places, but the inaccuracy is fairly small.  I think it's mostly related to the triangle loop now.

-[Unknown]

Reply:
Cool!

This immediate stuff is one pretty wacky feature. Wonder whether it was only intended for early hardware debugging or it was determined that it was confusing and not a performance win, and that's why it never was documented (presumably, due to the low usage).